fs/dafsclient/rmtclient.cpp

```diff
@@ -431,29 +435,23 @@ void receiveDaFsBuffer(ISocket * socket, MemoryBuffer & tgt, unsigned numtries,
             size32_t szread;
             SOCKREADTMS(socket)((gotLength<4000)?tgt.reserve(gotLength):tgt.reserveTruncate(gotLength), gotLength, gotLength, szread, timeout);
         }
-        catch (IJSOCK_Exception *e) {
-            if (e->errorCode()!=JSOCKERR_timeout_expired) {
-                EXCLOG(e,"receiveDaFsBuffer(1)");
-                PrintStackReport();
-                if (!tm.timemon||!tm.timemon->timedout())
-                    flushDaFsSocket(socket);
-            }
-            else {
-                EXCLOG(e,"receiveDaFsBuffer");
-                PrintStackReport();
-            }
+        catch (IException *e)
+        {
+            exception.setown(e);
             tgt.setLength(origlen);
-            throw;
         }
-        catch (IException *e) {
-            EXCLOG(e,"receiveDaFsBuffer(2)");
+
+        if (exception.get())
+        {
+            EXCLOG(exception, "receiveDaFsBuffer");
             PrintStackReport();
-            if (!tm.timemon||!tm.timemon->timedout())
-                flushDaFsSocket(socket);
-            tgt.setLength(origlen);
-            throw;
+            if (JSOCKERR_timeout_expired != exception->errorCode())
+            {
+                if (!tm.timemon||!tm.timemon->timedout())
+                    flushDaFsSocket(socket);
+            }
+            throw exception.getClear();
```

Comment:
@mckellyln Is there any potential issue if this was caught as a IJSOCK_Exception, but rethrown as an IException?  There are quite a few places that catch an IJSOCK_Exception then an IException and process them differently.

Reply:
Yes, there are two places where a catch from receiveDaFsBuffer() differentiates the type of exception.
I originally was planning to use two different exceptions and throw a matching type.  I'll update it now, thanks.

Reply:
@ghalliday updated to catch each exception type.  Is this what you meant.
Thanks for ffinding this.

Reply:
Right, with only a catch (IException *) and later throw, other catches will no longer see it as a IJSock_Exception (if that's what it was originally).
It's annoying and a problem elsewhere, where we generalize on catch (IException *).

Reply:
@mckellyln The problem isn't with the catch, the IException * will match both.

The problem is when the exception is thrown.  If throw; is used it is thrown with the type of the original exception.  If throw e is used it is thrown with the type that is thrown.  In this case it will be thrown with a type of IException *.

If there is some calling code that catches on (IJSOCK_Exception *) it will not match.  E.g. sendRemoteCommand() only has a 

catch (IJSOCK_Exception *e)

You could either have two owned variables, and throw the first that matches, or do a dynamic cast, and throw the IJSOCK_Exception * if the cast succeeds.

Reply:
I needed reminding about that problem too, thanks.